### PR TITLE
Allow and warn when indexing a zero-dim Variable

### PR DESCRIPTION
**Pull request by colesbury**
```
This better maintains backwards compatibility when Tensors and Variables
are merged. For example:

   >>> loss = var.sum().data[0]

Currently, `var.sum().data` is 1-dim so indexing. Once scalars are
enabled and Variable and Tensor are merged it will be zero-dim. This
change allows that expression to continue working (with a warning). In
the future, the canonical way to compute that expression will be:

   >>> loss = float(var.sum())

Or an equivalent alternative:

   >>> loss = var.sum().item()

Also fixes a few error cases.
```

### CODE REVIEW

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 1]
```diff
@@ -239,14 +249,16 @@ static PyObject* applyBoolGetitem(const Variable& self, bool index) {
...
+  AutoGPU auto_gpu(self_);
 
   // handle simple types: integers, slices, ellipsis
   if (index == Py_None) {
     return wrap(self_.unsqueeze(0));
   } else if (index == Py_Ellipsis) {
     return wrap(at::alias(self_));
   } else if (THPUtils_checkLong(index)) {
-    return wrap(applySelect(self_, 0, THPUtils_unpackLong(index)));
+    auto index_val = THPUtils_unpackLong(index);
```
Comment (gchanan) [inline torch/csrc/autograd/python_variable_indexing.cpp]: did you mean to leave this in?

**Review by gchanan — APPROVED** [review]
This looks good but can we merge this after we turn on WITH_SCALARS?
